src/ServiceControl.Monitoring/Infrastructure/EndpointRegistry.cs

```diff
@@ -1,30 +1,33 @@
 ﻿namespace ServiceControl.Monitoring.Infrastructure
 {
-    using System.Collections.Concurrent;
     using System.Collections.Generic;
     using System.Linq;
 
     public class EndpointRegistry
     {
-        ConcurrentDictionary<EndpointInstanceId, bool> endpointInstances = new ConcurrentDictionary<EndpointInstanceId, bool>();
+        HashSet<EndpointInstanceId> endpointInstances = new HashSet<EndpointInstanceId>();
+        volatile Dictionary<string, IEnumerable<EndpointInstanceId>> lookup = new Dictionary<string, IEnumerable<EndpointInstanceId>>();
+        object @lock = new object();
 
         public void Record(EndpointInstanceId endpointInstanceId)
         {
-            endpointInstances.TryAdd(endpointInstanceId, true);    
+            lock (@lock)
+            {
+                endpointInstances.Add(endpointInstanceId);
+                lookup = endpointInstances.ToArray()
+                    .GroupBy(instance => instance.EndpointName)
+                    .ToDictionary(g => g.Key, g => (IEnumerable<EndpointInstanceId>)g.Select(i => i).ToArray());
```

Comment:
Why do we convert to an array but cast as an IEnumerable? Should this be an IReadOnlyCollection maybe?